src/code/code_0x80186A70.c

```diff
@@ -0,0 +1,307 @@
+    func_800920B0(&D_801FD080, D_801FD080.baseAddress, &outFlashStatus);
+
+    *flash_status = outFlashStatus & 0xFF;
+
+    return;
+}
+
+void osFlashReadId(u32* flash_type, u32* flash_vendor) {
+    u8 flash_status; // sp1F
+
+    // why read status ?
+    osFlashReadStatus(&flash_status);
+
+    // select silicon id read mode
+    func_80093BB0(&D_801FD080, D_801FD080.baseAddress | FRAM_COMMAND_REGISTER, FRAM_COMMAND_SET_MODE_STATUS_AND_STATUS);
+
+    // read silicon id using DMA
+    D_801FD050.hdr.pri = 0;
+    D_801FD050.hdr.retQueue = &D_801FD068;
+    D_801FD050.dramAddr = D_801FD040;
+    D_801FD050.devAddr = 0;
+    D_801FD050.size = 8;
+
+    osInvalDCache(D_801FD040, 16);
+    osEPiStartDma(&D_801FD080, &D_801FD050, 0);
+    osRecvMesg(&D_801FD068, 0, 1);
+
+    *flash_type = D_801FD040[0];
+    *flash_vendor = D_801FD040[1];
+
+    return;
+}
+
+void osFlashClearStatus(void) {
+    // select status mode
+    func_80093BB0(&D_801FD080, D_801FD080.baseAddress | FRAM_COMMAND_REGISTER, FRAM_COMMAND_EXECUTE);
+    // clear status
+    func_80093BB0(&D_801FD080, D_801FD080.baseAddress, 0x00000000);
+
+    return;
+}
+
+s32 osFlashAllErase(void) {
+    u32 status;     // sp6C
+    OSTimer timer;  // sp48
+    OSMesgQueue mq; // sp30
+    OSMesg msg;     // sp2C
+
+    // start chip erase operation
+    func_80093BB0(&D_801FD080, D_801FD080.baseAddress | FRAM_COMMAND_REGISTER, FRAM_COMMAND_UNK_ERASE_OPERATION);
+    func_80093BB0(&D_801FD080, D_801FD080.baseAddress | FRAM_COMMAND_REGISTER, FRAM_COMMAND_SET_MODE_ERASE_AND_STATUS);
+
+    // wait for completion by polling erase-busy flag
+    osCreateMesgQueue(&mq, &msg, 1);
+    do {
+        osSetTimer(&timer, 0xABA95ULL, 0ULL, &mq, &msg);
+        osRecvMesg(&mq, &msg, 1);
+        func_800920B0(&D_801FD080, D_801FD080.baseAddress, &status);
+    } while ((status & 0x2) == 0x2);
+
+    // check erase operation status, clear status
+    func_800920B0(&D_801FD080, D_801FD080.baseAddress, &status);
+    osFlashClearStatus();
+
+    if (((status & 0xFF) == 0x08) || ((status & 0xFF) == 0x48) || ((status & 0x08) == 0x08)) {
+        return 0;
+    } else {
+        return -1;
+    }
+}
+
+void osFlashAllEraseThrough(void) {
+    // start chip erase operation
+    func_80093BB0(&D_801FD080, D_801FD080.baseAddress | FRAM_COMMAND_REGISTER, FRAM_COMMAND_UNK_ERASE_OPERATION);
+    func_80093BB0(&D_801FD080, D_801FD080.baseAddress | FRAM_COMMAND_REGISTER, FRAM_COMMAND_SET_MODE_ERASE_AND_STATUS);
+}
+
+s32 osFlashCheckEraseEnd(void) {
+    u8 status; // sp1F
+
+    // check if erase operation is completed
+    osFlashReadStatus(&status);
+    if ((status & 0x02) == 0x02) {
+        return 2; // busy
+    } else {
+        // check erase operation status, clear status
+        osFlashReadStatus(&status);
+    }
+    osFlashClearStatus();
+
+    if (((status & 0xFF) == 0x08) || ((status & 0xFF) == 0x48) || ((status & 0x08) == 0x08)) {
+        return 0;
+    } else {
+        return -1;
+    }
+}
```

Comment:
Seems like there are a few functions here that return:
- `-1` error
- `0` success
- `2` busy

Should this be an enum?